### PR TITLE
Implementation of MSC2314

### PR DESCRIPTION
**Pull request by hawkowl**
MSC2314 is still in proposal, but here's an implementation that I can maybe use to bootstrap rescuing my server :D

MSC: https://github.com/matrix-org/matrix-doc/pull/2314

### CODE REVIEW

File: sytest-blacklist [hunk 1]
```diff
@@ -29,3 +29,7 @@ Enabling an unknown default rule fails with 404
 
 # Blacklisted due to https://github.com/matrix-org/synapse/issues/1663
 New federated private chats get full presence information (SYN-115)
+
+# Blacklisted due to https://github.com/matrix-org/matrix-doc/pull/2314 removing
+# this requirement from the spec
+Inbound federation of state requires event_id as a mandatory paramater
```
Comment (richvdh) [inline sytest-blacklist]: probably a better fix is to remove the test from sytest

File: tests/unittest.py [hunk 2]
```diff
@@ -559,6 +565,59 @@ def attempt_wrong_password_login(self, username, password):
         self.render(request)
         self.assertEqual(channel.code, 403, channel.result)
 
+    def inject_room_member(self, room, user, membership, replaces_state=None):
```
Comment (richvdh) [inline tests/unittest.py]: docstring on this wouldn't go amiss

File: sytest-blacklist (hunk 1 above)
Comment (richvdh) [reply sytest-blacklist]: I guess this is debatable until the MSC lands. :man_shrugging:

Comment (hawkowl) [reply sytest-blacklist]: yeah the spec requires event_id as mandatory, after the MSC lands we should def remove from sytest

**Pull request merged**